--- a/app_dual_mode_switch.h
+++ b/app_dual_mode_switch.h
@@ -61,6 +61,9 @@ typedef enum {
 *************************************************************************************
 ********************************************************************************** */
 
+#ifdef __cplusplus
+extern "C" {
+#endif
 /*! *********************************************************************************
 * \brief  Runs the dual mode switch Idle Task
 *
@@ -91,4 +94,8 @@ uint32_t dm_switch_get15_4InitWakeUpTime(void);
 ********************************************************************************** */
 void dm_switch_init15_4AfterWakeUp(void);
 
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* _APP_DUAL_MODE_SWITCH_H_ */
