--- a/app_dual_mode_low_power.h
+++ b/app_dual_mode_low_power.h
@@ -43,6 +43,9 @@ typedef enum
 * Public prototypes
 *************************************************************************************
 ********************************************************************************** */
+#ifdef __cplusplus
+extern "C" {
+#endif
 
 void dm_lp_init(void);
 
@@ -50,6 +53,10 @@ void dm_lp_preSleep(void);
 
 void dm_lp_wakeup(void);
 
-void dm_lp_processEvent(void *pParam);
+void dm_lp_processEvent(void * pParam);
+
+#ifdef __cplusplus
+}
+#endif
 
 #endif /* _APP_DUAL_MODE_LOW_POWER_H_ */
