--- a/gpio_pins.h
+++ b/gpio_pins.h
@@ -76,6 +76,11 @@
 #define BOARD_USER_BUTTON3_GPIO_PIN          4U
 #define IOCON_USER_BUTTON3_PIN               BOARD_USER_BUTTON3_GPIO_PIN
 
+#define BOARD_USER_BUTTON4_GPIO GPIO
+#define BOARD_USER_BUTTON4_GPIO_PORT 0U
+#define BOARD_USER_BUTTON4_GPIO_PIN 1U
+#define IOCON_USER_BUTTON4_PIN BOARD_USER_BUTTON4_GPIO_PIN
+
 #endif /* OM15082 */
 
 #define IOCON_USER_BUTTON_MODE_FUNC          (0U)
