--- a/MemManager.h
+++ b/MemManager.h
@@ -24,6 +24,9 @@
 #include "EmbeddedTypes.h"
 #include "GenericList.h"
 
+#ifdef __cplusplus
+extern "C" {
+#endif
 
 /*! *********************************************************************************
 *************************************************************************************
@@ -220,4 +223,8 @@ extern __attribute__((weak)) uint32_t MEM_GetTimeStamp(void);
 
 #endif /*MEM_TRACKING*/
 
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* _MEM_MANAGER_H_ */ 

