--- a/OtaUtils.c
+++ b/OtaUtils.c
@@ -92,11 +92,11 @@ typedef uint32_t (*flash_GetDmaccStatus_t)(uint8_t *address);
 *************************************************************************************
 ************************************************************************************/
 
-const static efuse_LoadUniqueKey_t efuse_LoadUniqueKey   = (efuse_LoadUniqueKey_t) ROM_API_efuse_LoadUniqueKey;
-const static aesLoadKeyFromOTP_t aesLoadKeyFromOTP       = (aesLoadKeyFromOTP_t) ROM_API_aesLoadKeyFromOTP;
-const static crc_update_t crc_update                       = (crc_update_t)ROM_API_crc_update;
-const static boot_CheckVectorSum_t boot_CheckVectorSum     = (boot_CheckVectorSum_t)ROM_API_boot_CheckVectorSum;
-const static flash_GetDmaccStatus_t flash_GetDmaccStatus   = (flash_GetDmaccStatus_t) ROM_API_flash_GetDmaccStatus;
+static const efuse_LoadUniqueKey_t efuse_LoadUniqueKey   = (efuse_LoadUniqueKey_t) ROM_API_efuse_LoadUniqueKey;
+static const aesLoadKeyFromOTP_t aesLoadKeyFromOTP       = (aesLoadKeyFromOTP_t) ROM_API_aesLoadKeyFromOTP;
+static const crc_update_t crc_update                     = (crc_update_t)ROM_API_crc_update;
+static const boot_CheckVectorSum_t boot_CheckVectorSum   = (boot_CheckVectorSum_t)ROM_API_boot_CheckVectorSum;
+static const flash_GetDmaccStatus_t flash_GetDmaccStatus = (flash_GetDmaccStatus_t) ROM_API_flash_GetDmaccStatus;
 
 /******************************************************************************
 *******************************************************************************
