--- a/OtaUtils.h
+++ b/OtaUtils.h
@@ -77,6 +77,10 @@ typedef otaUtilsResult_t (*OtaUtils_ReadBytes)(uint16_t nbBytes,
 *************************************************************************************
 ********************************************************************************** */
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 /*! *********************************************************************************
 * \brief  Allows to read bytes from the Internal Flash
 *
@@ -185,6 +189,8 @@ uint32_t OtaUtils_ValidateImage(OtaUtils_ReadBytes pFunctionRead,
 ********************************************************************************** */
 otaUtilsResult_t OtaUtils_ReconstructRootCert(IMAGE_CERT_T *pCert, const psector_page_data_t* pPage0, const psector_page_data_t* pFlashPage);
 
-
+#ifdef __cplusplus
+}
+#endif
 
 #endif /* _OTA_UTILS_H_ */
\ No newline at end of file
