--- a/SecLib.h
+++ b/SecLib.h
@@ -174,6 +174,10 @@ typedef void(*secLibCallback_t)(computeDhKeyParam_t* pData);
 *************************************************************************************
 ************************************************************************************/
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 /*! *********************************************************************************
 * \brief  This function performs initialization of the cryptographic HW acceleration.
 *
@@ -999,5 +1003,10 @@ uint32_t u32Reverse(uint32_t u32InWord);
 void vSwipeEndian( AESSW_Block_u *puBlock,
                    tsReg128 *psReg,
                    bool_t bBlockToReg );
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* _SEC_LIB_H_ */
 
