--- a/sign_images.sh
+++ b/sign_images.sh
@@ -56,7 +56,14 @@ if [ "$#" -gt 0 ]; then
 		MIME_SET="$(file -ib $FILENAME)"
 
 		if [[ $MIME_SET == *"$MIME_PATTERN"* ]]; then
-			python3 $SIGNING_TOOL $FILENAME
+			OTA_ENABLED=0
+			APP_START=`arm-none-eabi-nm $FILENAME | grep -i m_app_start | cut -d " " -f 1`
+
+			if [ "$APP_START" -gt 0 ]; then
+				OTA_ENABLED=1
+			fi
+
+			python3 $SIGNING_TOOL -i="$OTA_ENABLED" $FILENAME
 			arm-none-eabi-objcopy -O binary $FILENAME $FILENAME.bin
 		fi
 	done
